# PRÁCTICA 1 - APRENDIZAJE SUPERVISADO - PREPROCESADO

## 0. Indice
0. [Indice](#0-indice)
1. [Importación de Librerias, Lectura y Configuración](#11-importacion-lectura-y-configuracion)
2. [Entendimiento del DataFrame](#12-entendimiento-del-dataframe)
3. [Limpieza de los datos](#13-limpieza-de-los-datos)
    1. [Valores Nulos Numéricos](#131-valores-nulos-numericos)
    2. [Tratamiento de Columnas Nominales](#132-tratamiento-de-columnas-nominales)
        - [Tratamiento con Mapeo](#1321-tratamiento-por-mapeo)
        - [Tratamiento con OneHot-Encoding](#1322-tratamiento-con-onehot-encoding)
    3. [Tratamiento de Columnas Booleanas](#133-tratamiento-de-columnas-booleanas)
    4. [Tratamiento de Columnas DataTime](#134-tratamiento-de-columna-datatime)
4. [Combinación de Celdas por Cliente](#14-combinacion-de-celdas-por-cliente)
5. [Normalización de los Datos](#15-normalizacion-de-los-datos)
6. [Selección de Características](#16-seleccion-de-caracteristicas)
    1. [Filtro](#161-filtro)
        - [Chi^2](#1611-chi2)
        - [Mutual Information](#1612-mutual-information)
        - [Análisis de Varianza](#1613-analisis-de-varianza)
    2. [Wrapper](#162-wrapper)
        - [RFE - Recursive Feature Eliminator](#1621-recursive-feature-eliminator-rfe)
        - [RFECV - Recursive Feature Eliminator With Cross Validation](#1622-recursive-deature-elimination-with-cross-validation-rfecv)
        - [SFS - Sequential Feature Selector](#1623-sequential-feature-selector-sfs)
    3. [Embebido](#163-embebido)
7. [Tratamiento de los datos Desbalanceados](#17-tratamiento-de-los-datos-desbalanceados)
8. [Separación en Entrenamiento y Test](#18-separacion-en-entrenamiento-y-test)

## 1.1. Importacion, Lectura y Configuracion
[Volver al Indice](#indice)

Aqui nos encargamos de importar las librerias para hacer el preprocesado.<br>También hemos tocado alguna configuración para facilitar el visualizado de los datos.

In [337]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv("./data/train_data_0.csv")
df_label = pd.read_csv("./data/train_labels_0.csv")

In [339]:
# enseñar todas las columnas / si quiero todas las filas max_rows
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

## 1.2. Entendimiento del DataFrame

Vamos a intentar entender en que consiste el dataframe que nos han dado. Lo primero que vamos a ver es el tamaño del data set y los 5 primeros valores para ver a los que nos estamos enfrentando.

In [340]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3318787 entries, 0 to 3318786
Columns: 190 entries, ID to Infraction_DNOU
dtypes: float64(182), int64(1), object(7)
memory usage: 4.7+ GB


In [341]:
df.describe()

Payment_6804  Infraction_CGP     Base_7744    Base_80863     Risk_1930  \
count  3.291306e+06    3.318787e+06  3.318787e+06  3.317538e+06  3.318787e+06   
mean   6.560186e-01    1.531595e-01  1.239752e-01  6.211500e-01  7.896297e-02   
std    2.448160e-01    2.702251e-01  2.117188e-01  4.017901e-01  2.267270e-01   
min   -4.589548e-01    5.957676e-09 -7.588799e+00  9.192280e-09  1.534223e-09   
25%    4.798390e-01    4.526718e-03  8.887562e-03  1.046842e-01  2.898670e-03   
50%    6.940529e-01    9.058164e-03  3.141897e-02  8.143328e-01  5.786503e-03   
75%    8.646218e-01    2.366754e-01  1.258517e-01  1.002415e+00  8.661354e-03   
max    1.010000e+00    5.389619e+00  1.324059e+00  1.010000e+00  3.256284e+00   

       Expenditure_JIG  Infraction_SNZ    Base_02683  Infraction_SBF  \
count     2.708814e+06    3.317538e+06  3.317538e+06   477408.000000   
mean      2.262983e-01    5.978101e-02  1.326030e-01        0.185303   
std       1.940680e-01    2.022845e-01  2.347359e-01        0.227967   
min      -6.271320e-01    1.305867e-09  6.285293e-09       -0.000454   
25%       1.273215e-01    2.873889e-03  5.232143e-03        0.037655   
50%       1.639502e-01    5.745213e-03  9.783271e-03        0.120739   
75%       2.587185e-01    8.615531e-03  1.557437e-01        0.251833   
max       5.482888e+00    7.510052e+00  1.571396e+00        4.191119   

       Infraction_ZWWJ  Infraction_QJJF    Base_76065  Infraction_EJZ  \
count     2.321974e+06     3.154027e+06  3.318787e+06    3.317538e+06   
mean      1.546620e-01     1.185211e-01  1.723391e-01    2.518419e-01   
std       2.128579e-01     2.215333e-01  2.218131e-01    2.426288e-01   
min       1.154550e-07     2.496933e-09  3.099332e-09    1.563241e-08   
25%       4.225719e-02     3.834216e-03  2.755648e-02    5.442996e-02   
50%       8.861963e-02     7.675366e-03  8.220385e-02    1.800831e-01   
75%       1.843370e-01     1.319647e-01  2.389794e-01    3.732903e-01   
max       1.011162e+01     5.634724e+00  1.980329e+01    1.601498e+00   

          Base_6872     Risk_0322  Infraction_FMXQ  Infraction_GGO  \
count  3.318787e+06  3.318787e+06     2.592299e+06    3.318787e+06   
mean   8.302400e-02  4.761441e-02     4.754497e-01    4.067701e-01   
std    3.622044e-01  2.020024e-01     1.681405e-01    2.347571e-01   
min    2.804822e-11  8.265748e-10    -5.427248e+00   -2.662164e-02   
25%    7.295088e-03  2.613773e-03     4.251922e-01    2.321449e-01   
50%    1.540386e-02  5.223714e-03     4.597557e-01    3.822210e-01   
75%    5.392817e-02  7.837555e-03     5.174431e-01    5.614324e-01   
max    6.093663e+01  1.010000e+00     9.834361e+00    1.643096e+00   

       Infraction_TLPJ  Infraction_NCB     Base_1165    Base_39598  \
count     2.888970e+06    3.289380e+05  3.318635e+06  3.318787e+06   
mean      3.822308e-01    1.915690e-01  1.523830e-01  1.862161e-01   
std       3.264468e-01    2.702727e-01  1.472253e+00  2.303733e-01   
min      -9.615277e-03    5.021298e-07 -5.075235e-03 -2.652748e+00   
25%       8.122885e-02    6.177662e-02  2.046470e-02  2.825864e-02   
50%       2.867120e-01    1.305211e-01  8.336161e-02  7.577154e-02   
75%       6.711215e-01    2.466849e-01  1.918492e-01  2.714976e-01   
max       5.106712e+00    4.584012e+01  1.214517e+03  1.252750e+00   

          Base_6187  Infraction_CLLY  Infraction_ZTNC    Base_85131  \
count  3.305442e+06     1.434881e+06     3.318787e+06  3.318787e+06   
mean   4.512175e-01     1.734578e-01     1.457189e-01  1.900599e-01   
std    4.970099e-01     6.581672e-01     2.430836e-01  2.868135e-01   
min    1.179499e-08    -3.064871e+02     3.351734e-09  6.726070e-09   
25%    4.514376e-03     6.475489e-02     3.649115e-03  5.762473e-03   
50%    9.029771e-03     1.092322e-01     7.292170e-03  2.609364e-02   
75%    1.004181e+00     1.864358e-01     3.360667e-01  3.357074e-01   
max    1.015023e+00     2.440427e+02     2.676606e+00  2.742486e+01   

          Risk_9995  Infraction_AYWV  Payment_22507     Base_9516  \

In [342]:
df.head()

ID Expenditure_AHF  \
0  3333337004103300505242400473433643475477705348...      2017-03-05   
1  3333337004103300505242400473433643475477705348...      2017-04-20   
2  3333337004103300505242400473433643475477705348...      2017-05-11   
3  3333337004103300505242400473433643475477705348...      2017-06-29   
4  3333337004103300505242400473433643475477705348...      2017-08-03   

   Payment_6804  Infraction_CGP  Base_7744  Base_80863  Risk_1930  \
0      0.929122        0.382790   0.025782    1.002647   0.005515   
1      0.918431        0.002224   0.006806    1.008186   0.003287   
2      0.922563        0.567403   0.033713    1.007497   0.005594   
3      0.919253        0.213193   0.031170    1.008534   0.008996   
4      0.901860        0.325467   0.109644    1.005098   0.008041   

   Expenditure_JIG  Infraction_SNZ  Base_02683  Infraction_SBF  \
0         0.089799        0.001479    0.005830             NaN   
1         0.095966        0.007831    0.000861             NaN   
2         0.094655        0.003162    0.010466             NaN   
3         0.099089        0.009513    0.005360             NaN   
4         0.108526        0.000732    0.005484             NaN   

   Infraction_ZWWJ  Infraction_QJJF  Base_76065  Infraction_EJZ  Base_6872  \
0              NaN         0.009873    0.021776        0.239459   0.070967   
1              NaN         0.007450    0.012991        0.247009   0.026492   
2              NaN         0.004221    0.015447        0.243067   0.031776   
3              NaN         0.001658    0.015686        0.254167   0.057811   
4              NaN         0.001454    0.046556        0.248585   0.029022   

   Risk_0322  Infraction_FMXQ  Infraction_GGO  Infraction_TLPJ  \
0   0.008058         0.457826        0.382562         0.059342   
1   0.006785         0.462680        0.386114         0.083383   
2   0.008117         0.450481        0.384930         0.086904   
3   0.007326         0.442022        0.389402         0.036248   
4   0.009446         0.432424        0.388776         0.031978   

   Infraction_NCB  Base_1165  Base_39598  Base_6187  Infraction_CLLY  \
0             NaN   0.201574    0.032390   0.007684              NaN   
1             NaN   0.202135    0.024356   0.003301              NaN   
2             NaN   0.191490    0.034360   0.006907              NaN   
3             NaN   0.199848    0.028317   0.008251              NaN   
4             NaN   0.226641    0.068204   0.000402              NaN   

   Infraction_ZTNC  Base_85131  Risk_9995  Infraction_AYWV  Payment_22507  \
0         0.334386    0.007561   0.005552         0.110471       0.619012   
1         0.672143    0.004459   0.001319         0.105637       0.591144   
2         0.666925    0.010246   0.002729         0.104565       0.597797   
3         0.340447    0.010398   0.008935         0.110119       0.610275   
4         0.341473    0.045093   0.002589         0.104974       0.566030   

   Base_9516  Infraction_GWL  Expenditure_YTR  Base_36384  Expenditure_FIP  \
0   0.302757             NaN         0.025802    0.011677         0.003243   
1   0.295964             NaN         0.004938    0.000887         0.006905   
2   0.296058             NaN         0.011832    0.017219         0.007552   
3   0.294000             NaN         0.036272    0.012104         0.000044   
4   0.301472             NaN         0.052949    0.081246         0.004738   

   Infraction_PAS  Risk_0003  Expenditure_HMO  Base_24406  Expenditure_LMSR  \
0        1.009099   0.001485         0.072583    0.026337          0.596242   
1        1.008566   0.000638         0.081383    0.019455          0.472848   
2        1.001835   0.004912         0.083990    0.022642          0.666716   
3        1.006224   0.003507         0.081560    0.025763          0.762724   
4        1.006219   0.000154         0.094930    0.022630          0.763803   

   Infraction_BSU  Infraction_GEL  Base_14808  Risk_8065  Infraction_ZYW  \
0        0.052813      

Aqui vamos a comprobar los datos que son del tipo objecto para ver luego que hacer con ellos.

In [343]:
df.describe(include='object')

ID Expenditure_AHF  \
count                                             3318787         3318787   
unique                                             275347             401   
top     3333410542856622046303630323044578355701380044...      2018-02-18   
freq                                                   13           11480   

       Infraction_YFSG Infraction_DQLY Infraction_CLH    Base_67254  \
count          3318787         3188123        3188733       3317538   
unique               6               4              7             7   
top                 CO               O      very_high  moderate_low   
freq           2473189         1748616        1667728       1174012   

       Infraction_TEN  
count         3212364  
unique              7  
top     extremely_low  
freq           871883

Nos damos cuenta de que tenemos 275347 clientes (cuando tenemos más de 1M de tuplas), y cada uno de estos clientes tienen muchas facturas

Ahora vamos a ver información más general del dataframe para hacernos a la idea de los valores que almacenan cada una de las columnas

In [344]:
df.dtypes

ID                   object
Expenditure_AHF      object
Payment_6804        float64
Infraction_CGP      float64
Base_7744           float64
Base_80863          float64
Risk_1930           float64
Expenditure_JIG     float64
Infraction_SNZ      float64
Base_02683          float64
Infraction_SBF      float64
Infraction_ZWWJ     float64
Infraction_QJJF     float64
Base_76065          float64
Infraction_EJZ      float64
Base_6872           float64
Risk_0322           float64
Infraction_FMXQ     float64
Infraction_GGO      float64
Infraction_TLPJ     float64
Infraction_NCB      float64
Base_1165           float64
Base_39598          float64
Base_6187           float64
Infraction_CLLY     float64
Infraction_ZTNC     float64
Base_85131          float64
Risk_9995           float64
Infraction_AYWV     float64
Payment_22507       float64
Base_9516           float64
Infraction_GWL      float64
Expenditure_YTR     float64
Base_36384          float64
Expenditure_FIP     float64
Infraction_PAS      

Aqui nos podemos fijar que la mayoría de los datos ya son valores numéricos, pero en vez de ser enteros son con coma (float). Luego hay un dato que es un entero de 64 bits y para acabar hay 7 que son del tipo objeto.

En los siguientes pasos he divido en los distintos tipos de caracteríticas que tengo para profundizar en el entendimiento del dataset.

In [345]:
# Seleccionar solo las columnas que contienen 'Infraction' en su nombre
df_infractions = df.filter(regex='Infraction')

# aqui muestro la dimension del dataframe para ver las columnas dedicadas a 
# 'infractions'
print(df_infractions.shape)

# Mostrar las primeras filas para verificar el resultado
df_infractions.head(10)

(3318787, 96)


Infraction_CGP  Infraction_SNZ  Infraction_SBF  Infraction_ZWWJ  \
0        0.382790        0.001479             NaN              NaN   
1        0.002224        0.007831             NaN              NaN   
2        0.567403        0.003162             NaN              NaN   
3        0.213193        0.009513             NaN              NaN   
4        0.325467        0.000732             NaN              NaN   
5        0.032285        0.005400             NaN              NaN   
6        0.004155        0.007793             NaN         0.525600   
7        0.038577        0.007951             NaN         0.125986   
8        0.003067        0.004865             NaN         0.093257   
9        0.302357        0.000728             NaN         0.079258   

   Infraction_QJJF  Infraction_EJZ  Infraction_FMXQ  Infraction_GGO  \
0         0.009873        0.239459         0.457826        0.382562   
1         0.007450        0.247009         0.462680        0.386114   
2         0.004221        0.243067         0.450481        0.384930   
3         0.001658        0.254167         0.442022        0.389402   
4         0.001454        0.248585         0.432424        0.388776   
5         0.004783        0.259658         0.436873        0.388474   
6         0.005822        0.261886         0.453021        0.390310   
7         0.004200        0.258513         0.444563        0.394144   
8         0.001759        0.258391         0.447928        0.393696   
9         0.002812        0.267228         0.471737        0.399547   

   Infraction_TLPJ  Infraction_NCB  Infraction_CLLY  Infraction_ZTNC  \
0         0.059342             NaN              NaN         0.334386   
1         0.083383             NaN              NaN         0.672143   
2         0.086904             NaN              NaN         0.666925   
3         0.036248             NaN              NaN         0.340447   
4         0.031978             NaN              NaN         0.341473   
5         0.030584             NaN              NaN         0.342604   
6         0.067774             NaN              NaN         0.343320   
7         0.060045             NaN              NaN         0.333742   
8         0.028050             NaN              NaN         0.341949   
9         0.105999             NaN              NaN         0.339235   

   Infraction_AYWV  Infraction_GWL  Infraction_PAS  Infraction_BSU  \
0         0.110471             NaN        1.009099        0.052813   
1         0.105637             NaN        1.008566        0.060770   
2         0.104565             NaN        1.001835        0.059962   
3         0.110119             NaN        1.006224        0.052496   
4         0.104974             NaN        1.006219        0.048872   
5         0.103495             NaN        1.008006        0.045859   
6         0.111520             NaN        1.009100        0.050620   
7         0.108757             NaN        1.009799        0.049228   
8         0.239194             NaN        1.006363        0.039019   
9         0.241398             NaN        1.003239        0.039435   

   Infraction_GEL  Infraction_ZYW  Infraction_HSSU  Infraction_EHZP  \
0        0.684371        0.007069         0.230700         0.375191   
1        0.689035        0.004691         0.231395         0.295413   
2        0.717729        0.006135         0.237276         0.446750   
3        0.717408        0.005296         0.232312         0.333445   
4        0.717988        0.007024         0.237602         0.857541   
5        0.726356        0.002644         0.231371         0.105557   
6        0.695724        0.000004         0.232739         0.111703   
7        0.697400        0.004996         0.232800         0.059118   
8        0.700900        0.005523         0.235148         0.077823   
9        0.693418        0.000803         0.237700         0.403688   

   Infraction_TBP  Infraction_PBC Infraction_YFSG Infraction_DQLY  \
0        0.063612        0.239244  

In [346]:
# Seleccionar solo las columnas que contienen 'Expenditure' en su nombre
df_expenditure = df.filter(regex='Expenditure')

# aqui muestro la dimension del dataframe para ver las columnas dedicadas a 
# 'expenditure'
print(df_expenditure.shape)

# Mostrar las primeras filas para verificar el resultado
df_expenditure.head()

(3318787, 22)


Expenditure_AHF  Expenditure_JIG  Expenditure_YTR  Expenditure_FIP  \
0      2017-03-05         0.089799         0.025802         0.003243   
1      2017-04-20         0.095966         0.004938         0.006905   
2      2017-05-11         0.094655         0.011832         0.007552   
3      2017-06-29         0.099089         0.036272         0.000044   
4      2017-08-03         0.108526         0.052949         0.004738   

   Expenditure_HMO  Expenditure_LMSR  Expenditure_KMW  Expenditure_RGD  \
0         0.072583          0.596242              NaN         0.281806   
1         0.081383          0.472848         0.006782         0.368255   
2         0.083990          0.666716              NaN         0.408506   
3         0.081560          0.762724              NaN         0.282240   
4         0.094930          0.763803         0.007432         0.209741   

   Expenditure_IDZ  Expenditure_BWX  Expenditure_UWVG  Expenditure_HRQ  \
0         0.175123         0.282619          0.200131         0.001228   
1         0.133366         0.006044          0.205070         0.003258   
2         0.181359         0.428290          0.208094         0.003000   
3         0.344811         0.282940          0.101021         0.007752   
4         0.175933         0.286403          0.100353         0.001394   

   Expenditure_XDD  Expenditure_HKXV  Expenditure_GCAO  Expenditure_GMC  \
0         0.000797          0.003757          0.002157         0.008184   
1         0.003771          0.001069          0.008078         0.005347   
2         0.002065          0.006893          0.000952         0.002703   
3         0.007113          0.009890          0.005835         0.008242   
4         0.009012          0.002780          0.006392         0.006672   

   Expenditure_UIWS  Expenditure_ONEG  Expenditure_MTRQ  Expenditure_LAHK  \
0          0.914921          0.135440          0.926104          0.976465   
1          0.886502          0.133348          0.894015          0.976638   
2          0.939467          0.133085          0.934993          0.978863   
3          0.913676          0.132812          0.918448          0.975971   
4          0.947045          0.136809          0.960283          0.974565   

   Expenditure_HPM  Expenditure_BEH  
0         0.006042              NaN  
1         0.005882         0.317246  
2         0.001987         0.005694  
3         0.007262         0.002922  
4         0.001130         0.341470

In [347]:
# Seleccionar solo las columnas que contienen 'Risk' en su nombre
df_risk = df.filter(regex='Risk')

# aqui muestro la dimension del dataframe para ver las columnas dedicadas a 
# 'Risk'
print(df_risk.shape)

# Mostrar las primeras filas para verificar el resultado
df_risk.head()


(3318787, 28)


Risk_1930  Risk_0322  Risk_9995  Risk_0003  Risk_8065  Risk_1475  \
0   0.005515   0.008058   0.005552   0.001485   0.004105   0.000396   
1   0.003287   0.006785   0.001319   0.000638   0.000791   0.004131   
2   0.005594   0.008117   0.002729   0.004912   0.005254   0.004799   
3   0.008996   0.007326   0.008935   0.003507   0.004414   0.001317   
4   0.008041   0.009446   0.002589   0.000154   0.007880   0.008271   

   Risk_7095  Risk_6346  Risk_5797  Risk_2102  Risk_4804  Risk_4553  \
0   0.000963   0.009420        NaN   0.000954   0.008326   1.002751   
1   0.001982   0.003919        NaN   0.004326   0.001358   1.006914   
2   0.003067   0.005586        NaN   0.007756   0.004264   1.005440   
3   0.009390   0.004440        NaN   0.008919   0.004406   1.008088   
4   0.005445   0.007108        NaN   0.004627   0.009922   1.002697   

   Risk_8742  Risk_4247  Risk_2380  Risk_0454  Risk_4160  Risk_3506  \
0   0.007203   0.001399   0.003912   0.004460   0.007083   0.004040   
1   0.001203   0.009384   0.009396   0.002898   0.007837   0.003902   
2   0.002608   0.005944   0.000371   0.002720   0.008268   0.001903   
3   0.006692   0.007051   0.006375   0.005186   0.003785   0.009991   
4   0.008732   0.000852   0.009575   0.004130   0.002932   0.001728   

   Risk_9367  Risk_9423  Risk_6977  Risk_6178  Risk_6197  Risk_8532  \
0   0.009864        NaN   0.003199   0.000482   0.004166   0.001500   
1   0.004189   0.000296   0.003957   0.000256   0.009724   0.009940   
2   0.001200   0.003717   0.009305   0.003114   0.001053   0.006077   
3   0.001468   0.003763   0.006923   0.002302   0.004728   0.000317   
4   0.006513   0.001468   0.005330   0.008732   0.002751   0.004381   

   Risk_9247  Risk_4561  Risk_8902  Risk_5270  
0   0.004736        NaN   1.003868   0.003198  
1   0.009356        NaN   1.001694   0.009974  
2   0.003896        NaN   1.002329   0.002501  
3   0.007030        NaN   1.009615   0.006614  
4   0.000662        NaN   1.005002   0.003082

In [348]:
# Seleccionar solo las columnas que contienen 'Base' en su nombre
df_base = df.filter(regex='Base')

# aqui muestro la dimension del dataframe para ver las columnas dedicadas a 
# 'base'
print(df_base.shape)

# Mostrar las primeras filas para verificar el resultado
df_base.iloc[:5, :10]

(3318787, 40)


Base_7744  Base_80863  Base_02683  Base_76065  Base_6872  Base_1165  \
0   0.025782    1.002647    0.005830    0.021776   0.070967   0.201574   
1   0.006806    1.008186    0.000861    0.012991   0.026492   0.202135   
2   0.033713    1.007497    0.010466    0.015447   0.031776   0.191490   
3   0.031170    1.008534    0.005360    0.015686   0.057811   0.199848   
4   0.109644    1.005098    0.005484    0.046556   0.029022   0.226641   

   Base_39598  Base_6187  Base_85131  Base_9516  
0    0.032390   0.007684    0.007561   0.302757  
1    0.024356   0.003301    0.004459   0.295964  
2    0.034360   0.006907    0.010246   0.296058  
3    0.028317   0.008251    0.010398   0.294000  
4    0.068204   0.000402    0.045093   0.301472

In [349]:
# Seleccionar solo las columnas que contienen 'Payment' en su nombre
df_payment = df.filter(regex='Payment')

# aqui muestro la dimension del dataframe para ver las columnas dedicadas a 
# 'payment'
print(df_payment.shape)

# Mostrar las primeras filas para verificar el resultado
df_payment.head()

(3318787, 3)


Payment_6804  Payment_22507  Payment_3207
0      0.929122       0.619012      0.004637
1      0.918431       0.591144      0.009077
2      0.922563       0.597797      0.003788
3      0.919253       0.610275      0.003013
4      0.901860       0.566030      0.008953

## 1.3. Limpieza de los Datos


### 1.3.1. Valores Nulos Numericos

Primero nos vamos a encargar de eliminar los nulos de las primeras. Vamos a mirar el porcentaje de nulos de cada una de las caracteristicas para decidir si eliminar la columna, eliminar el dato o completar el dato.

In [350]:
porcentaje_nulos = df.isnull().mean() * 100
porcentaje_nulos = porcentaje_nulos[porcentaje_nulos > 50]

todas las columnas que estan por encima del 80% de nulos las vamos a eliminar ya que no nos dan informacion completa

In [351]:
df = df.drop(columns=porcentaje_nulos.index)

In [352]:
df.isnull().sum()

ID                        0
Expenditure_AHF           0
Payment_6804          27481
Infraction_CGP            0
Base_7744                 0
Base_80863             1249
Risk_1930                 0
Expenditure_JIG      609973
Infraction_SNZ         1249
Base_02683             1249
Infraction_ZWWJ      996813
Infraction_QJJF      164760
Base_76065                0
Infraction_EJZ         1249
Base_6872                 0
Risk_0322                 0
Infraction_FMXQ      726488
Infraction_GGO            0
Infraction_TLPJ      429817
Base_1165               152
Base_39598                0
Base_6187             13345
Infraction_ZTNC           0
Base_85131                0
Risk_9995                 0
Infraction_AYWV       17583
Payment_22507        181540
Base_9516                 0
Expenditure_YTR           0
Base_36384                0
Expenditure_FIP           0
Infraction_PAS         1249
Risk_0003                 0
Expenditure_HMO      609973
Base_24406                0
Expenditure_LMSR    

In [353]:
# Calculamos el porcentaje de nulos en el DataFrame
porcentaje_nulos = df.isnull().sum().sum() / df.size * 100
print(f"Porcentaje de valores nulos: {porcentaje_nulos:.2f}%")

Porcentaje de valores nulos: 1.97%


Los siguientes nulos los vamos a rellenar con la media de cada columna

In [354]:
from sklearn.impute import SimpleImputer

In [355]:
df_completo = pd.DataFrame(df)

numerical_columns = df_completo.select_dtypes(include=['number']).columns
imputer = SimpleImputer(strategy='mean')
df_completo[numerical_columns] = imputer.fit_transform(
        df_completo[numerical_columns])

df = df_completo

In [356]:
# Seleccionar solo las columnas numéricas y contar los valores nulos
nulos_columnas_numericas = df_completo.select_dtypes(include=['number']).isnull().sum()
print(nulos_columnas_numericas)

Payment_6804        0
Infraction_CGP      0
Base_7744           0
Base_80863          0
Risk_1930           0
Expenditure_JIG     0
Infraction_SNZ      0
Base_02683          0
Infraction_ZWWJ     0
Infraction_QJJF     0
Base_76065          0
Infraction_EJZ      0
Base_6872           0
Risk_0322           0
Infraction_FMXQ     0
Infraction_GGO      0
Infraction_TLPJ     0
Base_1165           0
Base_39598          0
Base_6187           0
Infraction_ZTNC     0
Base_85131          0
Risk_9995           0
Infraction_AYWV     0
Payment_22507       0
Base_9516           0
Expenditure_YTR     0
Base_36384          0
Expenditure_FIP     0
Infraction_PAS      0
Risk_0003           0
Expenditure_HMO     0
Base_24406          0
Expenditure_LMSR    0
Infraction_BSU      0
Base_14808          0
Risk_8065           0
Infraction_ZYW      0
Base_1039           0
Infraction_HSSU     0
Infraction_EHZP     0
Infraction_TBP      0
Base_0580           0
Expenditure_RGD     0
Infraction_PBC      0
Infraction

En este punto tenemos todos lo valores numerico nulos ya completados con la media. Ahora nos vamos a encargar de transformar los datos 'objects' a numericos.

### 1.3.2. Tratamiento de Columnas Nominales


Vamos a ver los valores unicos de las columnas tipo 'object' para decidir que tipo de tratamiento realizar.

In [357]:
columnas_objeto = df.columns[df.dtypes == 'object']

for col in columnas_objeto:
    print(f'Columna {col} unique vals: {df[col].unique()}')
columnas_objeto

Columna ID unique vals: ['3333337004103300505242400473433643475477705348074613224505537552'
 '3333412724050477534287115505554450368233003786622422204286025570'
 '3333755354307801420025735367852078550468735254721045250448405357' ...
 '3335366323134113200123452532512700424352703404485704610610530335'
 '3333630088467273760335373870175567785475140446072301870340006310'
 '3333410542856622046303630323044578355701380044351404008503435437']
Columna Expenditure_AHF unique vals: ['2017-03-05' '2017-04-20' '2017-05-11' '2017-06-29' '2017-08-03'
 '2017-08-23' '2017-09-19' '2017-10-24' '2017-11-28' '2018-01-02'
 '2018-01-30' '2018-02-10' '2018-03-29' '2017-04-04' '2017-04-25'
 '2017-06-02' '2017-06-28' '2017-07-20' '2017-08-10' '2017-09-22'
 '2017-10-27' '2017-11-29' '2018-01-22' '2018-02-07' '2018-04-02'
 '2017-04-01' '2017-04-24' '2017-06-01' '2017-06-24' '2017-07-24'
 '2017-08-09' '2017-09-24' '2017-10-31' '2017-12-01' '2018-01-01'
 '2018-02-01' '2018-03-01' '2018-04-01' '2017-03-23' '2017-05-02

Index(['ID', 'Expenditure_AHF', 'Infraction_YFSG', 'Infraction_DQLY',
       'Infraction_CLH', 'Base_67254', 'Infraction_TEN'],
      dtype='object')

#### 1.3.2.1. Tratamiento por Mapeo

In [358]:
frequency_map = {
    'very_low': 0,
    'low': 1,
    'moderate_low': 2,
    'moderate': 3,
    'moderate_high': 4,
    'high': 5,
    'very_high': 6
}
# Crear el imputador con la estrategia de moda
imputer = SimpleImputer(strategy='most_frequent')
for col in ['Infraction_CLH', 'Base_67254', 'Infraction_TEN']:
    df[col] = df[col].map(frequency_map)
    df[col] = imputer.fit_transform(df[[col]])

In [359]:
# Verificación del mapeo: mostrar los valores únicos en cada columna para ver el efecto del mapeo
for col in ['Infraction_CLH', 'Base_67254', 'Infraction_TEN']:
    print(f"Valores únicos en {col} después del mapeo e imputación:\n", df[col].unique())
    print("\n")

# Verificación de la imputación: mostrar el conteo de valores nulos después de la imputación
print("Conteo de valores nulos por columna después de la imputación:\n", df.isnull().sum())

Valores únicos en Infraction_CLH después del mapeo e imputación:
 [6. 2. 3. 5. 4. 0. 1.]


Valores únicos en Base_67254 después del mapeo e imputación:
 [2. 1. 3. 5. 6. 4.]


Valores únicos en Infraction_TEN después del mapeo e imputación:
 [3. 6. 4. 2. 1. 5.]


Conteo de valores nulos por columna después de la imputación:
 ID                       0
Expenditure_AHF          0
Payment_6804             0
Infraction_CGP           0
Base_7744                0
Base_80863               0
Risk_1930                0
Expenditure_JIG          0
Infraction_SNZ           0
Base_02683               0
Infraction_ZWWJ          0
Infraction_QJJF          0
Base_76065               0
Infraction_EJZ           0
Base_6872                0
Risk_0322                0
Infraction_FMXQ          0
Infraction_GGO           0
Infraction_TLPJ          0
Base_1165                0
Base_39598               0
Base_6187                0
Infraction_ZTNC          0
Base_85131               0
Risk_9995                0

#### 1.3.2.2. Tratamiento con OneHot-Encoding

In [360]:
df['Infraction_DQLY'] = df['Infraction_DQLY'].fillna(
    df['Infraction_DQLY'].mode())
columns_to_encode = ['Infraction_YFSG', 'Infraction_DQLY']
df = pd.get_dummies(df, columns=columns_to_encode)

In [361]:
# Listar las columnas que contienen 'Infraction_YFSG'
columnas_infraction_yfsg = [col for col in df.columns if 'Infraction_YFSG' in col]
print("Columnas de 'Infraction_YFSG':\n", columnas_infraction_yfsg)

# Listar las columnas que contienen 'Infraction_DQLY'
columnas_infraction_dqly = [col for col in df.columns if 'Infraction_DQLY' in col]
print("\nColumnas de 'Infraction_DQLY':\n", columnas_infraction_dqly)

Columnas de 'Infraction_YFSG':
 ['Infraction_YFSG_CL', 'Infraction_YFSG_CO', 'Infraction_YFSG_CR', 'Infraction_YFSG_XL', 'Infraction_YFSG_XM', 'Infraction_YFSG_XZ']

Columnas de 'Infraction_DQLY':
 ['Infraction_DQLY_-1', 'Infraction_DQLY_O', 'Infraction_DQLY_R', 'Infraction_DQLY_U']


In [362]:
df.dtypes

ID                     object
Expenditure_AHF        object
Payment_6804          float64
Infraction_CGP        float64
Base_7744             float64
Base_80863            float64
Risk_1930             float64
Expenditure_JIG       float64
Infraction_SNZ        float64
Base_02683            float64
Infraction_ZWWJ       float64
Infraction_QJJF       float64
Base_76065            float64
Infraction_EJZ        float64
Base_6872             float64
Risk_0322             float64
Infraction_FMXQ       float64
Infraction_GGO        float64
Infraction_TLPJ       float64
Base_1165             float64
Base_39598            float64
Base_6187             float64
Infraction_ZTNC       float64
Base_85131            float64
Risk_9995             float64
Infraction_AYWV       float64
Payment_22507         float64
Base_9516             float64
Expenditure_YTR       float64
Base_36384            float64
Expenditure_FIP       float64
Infraction_PAS        float64
Risk_0003             float64
Expenditur

### 1.3.3. Tratamiento de Columnas Booleanas


In [363]:
df.isnull().sum()

ID                    0
Expenditure_AHF       0
Payment_6804          0
Infraction_CGP        0
Base_7744             0
Base_80863            0
Risk_1930             0
Expenditure_JIG       0
Infraction_SNZ        0
Base_02683            0
Infraction_ZWWJ       0
Infraction_QJJF       0
Base_76065            0
Infraction_EJZ        0
Base_6872             0
Risk_0322             0
Infraction_FMXQ       0
Infraction_GGO        0
Infraction_TLPJ       0
Base_1165             0
Base_39598            0
Base_6187             0
Infraction_ZTNC       0
Base_85131            0
Risk_9995             0
Infraction_AYWV       0
Payment_22507         0
Base_9516             0
Expenditure_YTR       0
Base_36384            0
Expenditure_FIP       0
Infraction_PAS        0
Risk_0003             0
Expenditure_HMO       0
Base_24406            0
Expenditure_LMSR      0
Infraction_BSU        0
Base_14808            0
Risk_8065             0
Infraction_ZYW        0
Base_1039             0
Infraction_HSSU 

In [364]:
boolean_columns = df.select_dtypes('bool').columns
for col in boolean_columns:
        df[col] = df[col].fillna(
            df[col].value_counts().head(1))
        df[col] = df[col].astype(int)

In [365]:
df.dtypes

ID                     object
Expenditure_AHF        object
Payment_6804          float64
Infraction_CGP        float64
Base_7744             float64
Base_80863            float64
Risk_1930             float64
Expenditure_JIG       float64
Infraction_SNZ        float64
Base_02683            float64
Infraction_ZWWJ       float64
Infraction_QJJF       float64
Base_76065            float64
Infraction_EJZ        float64
Base_6872             float64
Risk_0322             float64
Infraction_FMXQ       float64
Infraction_GGO        float64
Infraction_TLPJ       float64
Base_1165             float64
Base_39598            float64
Base_6187             float64
Infraction_ZTNC       float64
Base_85131            float64
Risk_9995             float64
Infraction_AYWV       float64
Payment_22507         float64
Base_9516             float64
Expenditure_YTR       float64
Base_36384            float64
Expenditure_FIP       float64
Infraction_PAS        float64
Risk_0003             float64
Expenditur

### 1.3.4. Tratamiento de Columna DataTime

In [366]:
df['Expenditure_AHF'] = pd.to_datetime(df['Expenditure_AHF'])
df['Expenditure_AHF'] = df['Expenditure_AHF'].astype(int) / 10**9

In [367]:
df.dtypes

ID                     object
Expenditure_AHF       float64
Payment_6804          float64
Infraction_CGP        float64
Base_7744             float64
Base_80863            float64
Risk_1930             float64
Expenditure_JIG       float64
Infraction_SNZ        float64
Base_02683            float64
Infraction_ZWWJ       float64
Infraction_QJJF       float64
Base_76065            float64
Infraction_EJZ        float64
Base_6872             float64
Risk_0322             float64
Infraction_FMXQ       float64
Infraction_GGO        float64
Infraction_TLPJ       float64
Base_1165             float64
Base_39598            float64
Base_6187             float64
Infraction_ZTNC       float64
Base_85131            float64
Risk_9995             float64
Infraction_AYWV       float64
Payment_22507         float64
Base_9516             float64
Expenditure_YTR       float64
Base_36384            float64
Expenditure_FIP       float64
Infraction_PAS        float64
Risk_0003             float64
Expenditur

## 1.4. Combinacion de celdas por cliente

Vamos a agrupar las filas por ID de cliente y sumando las otras caracteristicas.

In [368]:
df = df.groupby('ID').sum()

## 1.5. Normalizacion de los Datos

Ahora vamos a normalizar los datos de tipo numéricos para que el entrenamiento sea mas eficiente y equititativo, y para que le cueste menos al modelo realizar predicciones.

In [369]:
from sklearn.preprocessing import StandardScaler

In [370]:
scaler = StandardScaler()
numerical_columns = df.select_dtypes(include=['number']).columns
df[numerical_columns] = scaler.fit_transform(
    df[numerical_columns])


In [371]:
# Verificar la media y desviación estándar de las columnas numéricas después de la estandarización
df[numerical_columns].describe()

Expenditure_AHF  Payment_6804  Infraction_CGP     Base_7744  \
count     2.753470e+05  2.753470e+05    2.753470e+05  2.753470e+05   
mean     -4.515937e-16  3.306698e-16    1.889726e-16  3.628878e-17   
std       1.000002e+00  1.000002e+00    1.000002e+00  1.000002e+00   
min      -4.247129e+00 -3.367218e+00   -9.321812e-01 -3.756808e+00   
25%       3.610622e-01 -7.735579e-01   -8.288780e-01 -5.422193e-01   
50%       3.622975e-01  1.437470e-01   -2.874439e-01 -4.291943e-01   
75%       3.632240e-01  8.914232e-01    5.087005e-01  1.474896e-02   
max       3.674373e-01  1.493842e+00    2.690722e+01  6.503899e+00   

         Base_80863     Risk_1930  Expenditure_JIG  Infraction_SNZ  \
count  2.753470e+05  2.753470e+05     2.753470e+05    2.753470e+05   
mean  -7.719672e-17  1.308912e-16    -2.306999e-17    2.345707e-17   
std    1.000002e+00  1.000002e+00     1.000002e+00    1.000002e+00   
min   -1.593739e+00 -4.990241e-01    -3.573367e+00   -3.471005e-01   
25%   -1.074819e+00 -4.673030e-01    -5.582337e-01   -3.196483e-01   
50%    4.379347e-01 -4.601158e-01    -1.771391e-01   -3.145946e-01   
75%    8.641593e-01  4.125850e-02     1.296030e-01   -3.056233e-01   
max    1.196130e+00  1.048304e+01     2.214840e+01    3.554576e+01   

         Base_02683  Infraction_ZWWJ  Infraction_QJJF    Base_76065  \
count  2.753470e+05     2.753470e+05     2.753470e+05  2.753470e+05   
mean   3.979186e-17    -3.526560e-16    -1.972561e-16 -1.483034e-16   
std    1.000002e+00     1.000002e+00     1.000002e+00  1.000002e+00   
min   -5.815509e-01    -1.022170e+00    -5.675967e-01 -7.784297e-01   
25%   -5.556120e-01    -5.361382e-01    -5.418800e-01 -6.359606e-01   
50%   -5.296604e-01    -1.437407e-01    -4.910994e-01 -4.156899e-01   
75%    1.154991e-01     8.032200e-02     9.971017e-02  2.740985e-01   
max    6.046493e+00     4.324386e+01     2.577930e+01  1.483286e+01   

       Infraction_EJZ     Base_6872     Risk_0322  Infraction_FMXQ  \
count    2.753470e+05  2.753470e+05  2.753470e+05     2.753470e+05   
mean    -1.347298e-16  3.132770e-17 -5.460413e-17    -1.059826e-15   
std      1.000002e+00  1.000002e+00  1.000002e+00     1.000002e+00   
min     -9.762567e-01 -2.776937e-01 -3.449289e-01    -2.313228e+01   
25%     -8.020048e-01 -2.498710e-01 -3.114015e-01    -1.087390e-01   
50%     -3.322840e-01 -2.071348e-01 -3.055215e-01     1.517037e-01   
75%      4.855719e-01 -5.241301e-02 -2.989865e-01     2.933920e-01   
max      5.663859e+00  1.426400e+02  7.528029e+00     2.976679e+01   

       Infraction_GGO  Infraction_TLPJ     Base_1165    Base_39598  \
count    2.753470e+05     2.753470e+05  2.753470e+05  2.753470e+05   
mean    -4.003443e-16     2.977938e-17 -8.670600e-18 -2.262872e-16   
std      1.000002e+00     1.000002e+00  1.000002e+00  1.000002e+00   
min     -1.620061e+00    -1.312767e+00 -1.556985e-01 -1.664638e+00   
25%     -7.626940e-01    -8.605938e-01 -1.296616e-01 -6.864220e-01   
50%     -1.073705e-01    -2.251171e-01 -7.516893e-02 -4.845868e-01   
75%      6.743873e-01     6.855235e-01  5.274840e-02  3.786453e-01   
max      4.126516e+00     1.178351e+01  3.678541e+02  5.242940e+00   

          Base_6187  Infraction_ZTNC    Base_85131     Risk_9995  \
count  2.753470e+05     2.753470e+05  2.753470e+05  2.753470e+05   
mean  -8.071915e-17    -1.530258e-17  4.784313e-17 -1.318654e-16   
std    1.000002e+00     1.000002e+00  1.000002e+00  1.000002e+00   
min   -9.284144e-01    -6.257616e-01 -7.218264e-01 -7.078865e-01   
25%   -9.172984e-01    -6.041104e-01 -6.990795e-01 -6.681812e-01   
50%   -5.857255e-01    -5.980826e-01 -5.320550e-01 -3.241511e-01   
75%    1.299778e+00     4.661992e-01  5.110253e-01  2.294860e-01   
max    1.309678e+00     1.115796e+01  5.174404e+01  4.367819e+01   

       Infraction_AYWV  Payment_22507     Base_9516  Expenditure_YTR  \
count     2.753470e+05   2.753470e+05  2.753470e+05     2.753470e+05   
mean      2.052816e-16   7.994499e-16 -3.819193e-17     6.022970e-17   
std       1.00

Esto nos permitirá ver que las columnas ahora tienen una media cercana a 0 y una desviación estándar cercana a 1 y saber que hemos realizado bien la normalizacion

## 1.6. Seleccion de Caracteristicas

### 1.6.1. Filtro

In [372]:
from sklearn.feature_selection import SelectKBest, chi2, mutual_info_classif, f_classif, VarianceThreshold

X = df
y = df_label
y = y.drop('ID', axis=1)

#### 1.6.1.1. Chi2

In [373]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
X_scaled = pd.DataFrame(scaler.fit_transform(X), columns=X.columns)

selector = SelectKBest(score_func=chi2, k=50)
X_chi2 = selector.fit_transform(X_scaled, y)

chi2_selected_columns = X_scaled.columns[selector.get_support()]
print(chi2_selected_columns)

Index(['Risk_1930', 'Base_02683', 'Expenditure_FIP', 'Expenditure_LMSR',
       'Infraction_ZYW', 'Infraction_HSSU', 'Base_91828', 'Infraction_ZTYG',
       'Risk_2102', 'Risk_4247', 'Risk_2380', 'Infraction_MHM', 'Risk_9367',
       'Expenditure_GMC', 'Risk_6977', 'Risk_6178', 'Infraction_WVC',
       'Infraction_QVSL', 'Infraction_QXUM', 'Risk_8532', 'Infraction_IMIM',
       'Infraction_LTIS', 'Infraction_HFSI', 'Infraction_ETH', 'Base_67254',
       'Infraction_NMCB', 'Infraction_TPAF', 'Infraction_ZRH',
       'Infraction_XEPQ', 'Infraction_ZMKI', 'Infraction_TFOY',
       'Infraction_WMAQ', 'Infraction_CZXL', 'Base_52892', 'Infraction_HUK',
       'Infraction_QWWW', 'Infraction_YQXM', 'Infraction_QGR',
       'Infraction_LSX', 'Infraction_IBJ', 'Infraction_DNOU',
       'Infraction_YFSG_CL', 'Infraction_YFSG_CO', 'Infraction_YFSG_CR',
       'Infraction_YFSG_XL', 'Infraction_YFSG_XM', 'Infraction_YFSG_XZ',
       'Infraction_DQLY_-1', 'Infraction_DQLY_O', 'Infraction_DQLY_R'],
  

#### 1.6.1.2. Mutual Information

In [374]:
selector = SelectKBest(score_func=mutual_info_classif, k=50)
X_mut_info = selector.fit_transform(X, y)

mut_info_selected_columns = X.columns[selector.get_support()]
print(mut_info_selected_columns)

C:\Users\peric\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:1339: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Index(['Expenditure_AHF', 'Payment_6804', 'Base_7744', 'Infraction_QJJF',
       'Base_6872', 'Infraction_FMXQ', 'Infraction_GGO', 'Infraction_TLPJ',
       'Base_85131', 'Payment_22507', 'Expenditure_FIP', 'Infraction_PAS',
       'Base_1039', 'Infraction_PBC', 'Base_91828', 'Infraction_CLH',
       'Infraction_EYU', 'Payment_3207', 'Base_8730', 'Risk_4160', 'Risk_3506',
       'Base_23737', 'Risk_9423', 'Risk_6977', 'Risk_8532', 'Expenditure_ONEG',
       'Expenditure_LAHK', 'Infraction_LTIS', 'Base_67254', 'Infraction_NMCB',
       'Infraction_ZRH', 'Infraction_TEN', 'Infraction_XEPQ', 'Infraction_WIS',
       'Infraction_RKTA', 'Infraction_WMAQ', 'Infraction_CZXL', 'Risk_5270',
       'Infraction_QWWW', 'Infraction_YQXM', 'Infraction_YFSG_CL',
       'Infraction_YFSG_CO', 'Infraction_YFSG_CR', 'Infraction_YFSG_XL',
       'Infraction_YFSG_XM', 'Infraction_YFSG_XZ', 'Infraction_DQLY_-1',
       'Infraction_DQLY_O', 'Infraction_DQLY_R', 'Infraction_DQLY_U'],
      dtype='object')


#### 1.6.1.3. Analisis de Varianza

In [375]:
selector = SelectKBest(score_func=f_classif, k=50)
X_f_classif = selector.fit_transform(X, y)

f_classif_selected_columns = X.columns[selector.get_support()]
print(f_classif_selected_columns)

C:\Users\peric\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:1339: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Index(['Infraction_CGP', 'Risk_1930', 'Expenditure_JIG', 'Infraction_FMXQ',
       'Base_1165', 'Payment_22507', 'Base_9516', 'Expenditure_FIP',
       'Expenditure_HMO', 'Expenditure_LMSR', 'Base_14808', 'Infraction_ZYW',
       'Infraction_HSSU', 'Infraction_CLH', 'Base_8511', 'Infraction_ZTYG',
       'Infraction_QKZN', 'Infraction_CZE', 'Risk_7095', 'Infraction_JBR',
       'Risk_2102', 'Risk_4247', 'Risk_2380', 'Infraction_MHM', 'Risk_6977',
       'Risk_6178', 'Infraction_WVC', 'Expenditure_UIWS', 'Expenditure_MTRQ',
       'Expenditure_HPM', 'Infraction_LTIS', 'Base_5441', 'Base_67254',
       'Infraction_TPAF', 'Infraction_ZRH', 'Infraction_XEPQ',
       'Infraction_ZMKI', 'Infraction_WVAW', 'Infraction_TFOY',
       'Infraction_WMAQ', 'Base_52892', 'Risk_5270', 'Infraction_YQXM',
       'Infraction_IBJ', 'Infraction_DNOU', 'Infraction_YFSG_CL',
       'Infraction_YFSG_XL', 'Infraction_YFSG_XM', 'Infraction_DQLY_-1',
       'Infraction_DQLY_R'],
      dtype='object')


### 1.6.2. Wrapper

In [ ]:
from sklearn.feature_selection import RFE, RFECV, SequentialFeatureSelector
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

#### 1.6.2.1. Recursive Feature Eliminator (RFE)

In [ ]:

model = RandomForestClassifier()
selector = RFE(estimator=model, n_features_to_select=50)

X_rfe = selector.fit_transform(X, y)

selected_columns = X.columns[selector.get_support()]
print(selected_columns)

C:\Users\peric\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\base.py:1473: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)
C:\Users\peric\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\base.py:1473: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)
C:\Users\peric\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\base.py:1473: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y 

#### 1.6.2.2. Recursive Deature Elimination with Cross-Validation (RFECV)

In [ ]:
model = RandomForestClassifier()
selector = RFECV(estimator=model, cv=5) 
X_rfecv = selector.fit_transform(X, y)

selected_columns = X.columns[selector.get_support()]
print(selected_columns)

#### 1.6.2.3. Sequential Feature Selector (SFS)

In [ ]:
model = LogisticRegression()
selector = SequentialFeatureSelector(model, n_features_to_select=50, direction='forward', cv=5)
X_sfs = selector.fit_transform(X, y)

# Obtener nombres de las columnas seleccionadas
selected_columns = X.columns[selector.get_support()]
print(selected_columns)

### 1.6.3. Embebido

## 1.7. Tratamiento de los datos Desbalanceados

## 1.8. Separacion en Entrenamiento y Test

---
# 
Este Práctica se ha realizado por el Grupo_26 -> Pedro Llinas Ferrer y Andrés Benito Llorente